core/new-gui/src/app/workspace/service/workflow-graph/model/sync-texera-model.ts

```diff
@@ -0,0 +1,154 @@
+   *  - When the user drags the link from one port, but not yet to connect to another port,
+   *      the link is added in the semantic of JointJS, but not in the semantic of Texera Workflow graph.
+   *  - When an invalid link that is not connected to a port disappears,
+   *      the link delete event is trigger by JointJS, but the link never existed from Texera's perspective.
+   *  - When the user drags and detaches the end of a valid link, the link is disconnected from the target port,
+   *      the link change event (not delete) is triggered by JointJS, but the link should be deleted from Texera's graph.
+   *  - When the user attaches the end of the link to a target port,
+   *      the link change event (not add) is triggered by JointJS, but it should be added to the Texera Graph.
+   *  - When the user drags the link around, the link change event will be trigger continuously,
+   *      when the target being a changing coordinate. But this event should not have any effect on the Texera Graph.
+   *
+   * To address the disparity of the semantics, the linkAdded / linkDeleted / linkChanged events need to be handled carefully.
+   */
+  private handleJointLinkEvents(): void {
+    /**
+     * on link cell add:
+     * we need to check if the link is a valid link in Texera's semantic (has both source and target port)
+     *  and only add valid links to the graph
+     */
+    this.jointGraphWrapper.getJointLinkCellAddStream()
+      .filter(link => SyncTexeraModel.isValidJointLink(link))
+      .map(link => SyncTexeraModel.getOperatorLink(link))
+      .subscribe(link => this.texeraGraph.addLink(link));
+
+    /**
+     * on link cell delete:
+     * we need to first check if the link is a valid link
+     *  then delete the link by the link ID
+     */
+    this.jointGraphWrapper.getJointLinkCellDeleteStream()
+      .filter(link => SyncTexeraModel.isValidJointLink(link))
+      .subscribe(link => this.texeraGraph.deleteLinkWithID(link.id.toString()));
+
+
+    /**
+     * on link cell change:
+     * link cell change could cause deletion of a link or addition of a link, or simply no effect
+     * TODO: finish this documentation
+     */
+    this.jointGraphWrapper.getJointLinkCellChangeStream()
+      // we intentially want the side effect (delete the link) to happen **before** other operations in the chain
+      .do((link) => {
+        const linkID = link.id.toString();
+        if (this.texeraGraph.hasLinkWithID(linkID)) { this.texeraGraph.deleteLinkWithID(linkID); }
+      })
+      .filter(link => SyncTexeraModel.isValidJointLink(link))
+      .map(link => SyncTexeraModel.getOperatorLink(link))
+      .subscribe(link => {
+        this.texeraGraph.addLink(link);
+      });
+  }
+
+  /**
+   * Transforms a JointJS link (joint.dia.Link) to a Texera Link object
+   * The JointJS link must be valid, otherwise an error will be thrown.
+   * @param jointLink
+   */
+  static getOperatorLink(jointLink: joint.dia.Link): OperatorLink {
+
```

Comment:
rename to jointLinkEndpointType or something similar

Reply:
fixed